### ISSUE TITLE
Query for checking that K8s prometheus integration has begun sending data

### ISSUE THREAD

**Issue by nickbp** [opstrace/opstrace#690]
We'd like to show an indicator in the UI that we have started receiving prometheus metrics for a given integration. The expectation is that the relabeling config included in the integration will inject metrics with an integration UUID that we can then check for when querying cortex.

At the moment the plan is to keep things simple and just query via the grafana API as we do for checking exporters. Something to figure out is a good "canary" metric to query for, or alternatively just check for "recently received any metric with integration_id=x".

Depends on #687

**Comment by nickbp** [comment]
While working on 687 it looks like `apiserver_request_total` might be a good metric to query against the matching `integration_id` label..

**Comment by nickbp** [comment]
A catch with that `apiserver_request_total` metric is that it's marked as `ALPHA` as of 1.20, which doesn't give me a lot of confidence that it will be available in older K8s versions:
```
# HELP apiserver_request_total [ALPHA] Counter of apiserver requests broken out for each verb, dry run value, group, version, resource, scope, component, and HTTP response contentType and code.
# TYPE apiserver_request_total counter
```

In fact if I `grep -v ALPHA` the metrics docs, all of the K8s ones seem to be flagged `ALPHA`, leaving only the underlying golang ones like `go_info` or `go_threads`:
```
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.15.10"} 1
[...]
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 38
```

If those have existed for a long time across golang versions, they might be a safe option as a "canary metric".

Alternatively, if we are able to query cortex for _any_ metric with `integration_id=foo`, then that could be a very safe option - but I could see it being an expensive query if things are primarily indexed by metric name.

**Comment by nickbp** [comment]
Tried querying for exactly `{integration_id="foo"}` (no metric name with only the test integration ID) and it indeed behaved poorly. The Grafana UI choked and eventually displayed a `Request Entity Too Large` error.

Meanwhile querying `go_info{integration_id="foo"}` returned nearly instantaneously.

So it does look like we'll want to pick a metric that's been around for a while in K8s responses.

**Comment by nickbp** [comment]
Did some digging through K8s system component code. It looks like they have a [common library](https://github.com/kubernetes/component-base/blob/e913a97091533a577e45a7cba5f66b540366548c/metrics/legacyregistry/registry.go#L48) for their metrics, that then includes a couple default collectors from the prometheus go client going back to [Dec 2014](https://github.com/prometheus/client_golang/commit/dbd48d666b7cab8fdc1c5fea89ad1db67f79ad03):

- [`go_collector.go`](https://github.com/prometheus/client_golang/blob/master/prometheus/go_collector.go): Has a couple metrics largely unchanged back to 2015: [`go_gc_duration_seconds`](https://github.com/prometheus/client_golang/commit/ff586eaac16c26ec85429cc0b8c0ab72f1c6e2c5) and [`go_goroutines`](https://github.com/prometheus/client_golang/commit/765c1152ca986104b1d019bfe13ac3bcbb3e4b5a) (briefly `process_goroutines`).
- [`process_collector.go`](https://github.com/prometheus/client_golang/blob/master/prometheus/process_collector.go): Has [several metrics](https://github.com/prometheus/client_golang/commit/dbd48d666b7cab8fdc1c5fea89ad1db67f79ad03#diff-96f5c93065b3b2e660d109a58912dcc50d8e5fe16615152d636ae9c0d4db4ffeR35) largely unchanged back to the introduction of these utilities at the end of 2014. Among these I think `process_start_time_seconds` or `process_cpu_seconds_total` are the best bet since I could see them being widely used (and thereby effectively forced into being relatively stable for now on). I'd lean towards `process_cpu_seconds_total` since it'd be expected to at least slowly change over time, whereas `process_start_time_seconds` would mainly tell us if processes are restarting/crashing.

So in summary I'd say `process_cpu_seconds_total{integration_id="foo"}` would be a good canary for detecting that metrics have started arriving. Anything golang-based (particularly any K8s components) is pretty likely to have this metric since it dates back to Dec 2014 in the golang Prometheus library.

**Comment by nickbp** [comment]
I successfully deployed the prometheus configuration into the EKS cluster used by a test opstrace instance (so snake eating own tail in this case) and it was also able to talk to nodes, apiserver, and the subset of pods that had a `/metrics` endpoint going. Similarly, the `process_cpu_seconds_total{integration_id="foo"}` metric was returned by most of them (lot of go code in there). So we've at least got a sample size of two clusters where the config works, and where this metric works as a "canary".

If the cluster is very large and we want to narrow down the metric a bit, we could add a `job="default/kubernetes"` filter to the query to only return the apiserver metric while omitting node/pod metrics. But at the same time, that comes at the cost of thinking that we aren't getting anything if the apiserver isn't working but the nodes/pods are working.

**Comment by nickbp** [comment]
Some example instant/ranged queries against the metric that should be usable via the browser (with `anothernickstestid20210510` as the `integration_id`):
- `https://default.nick-test.opstrace.io/grafana/api/datasources/proxy/1/api/v1/query?query=process_cpu_seconds_total{integration_id%3D%22anothernickstestid20210510%22}&time=2021-05-10T01:43:42Z`
- `https://default.MYCLUSTER.opstrace.io/grafana/api/datasources/proxy/1/api/v1/query_range?query=process_cpu_seconds_total{integration_id%3D%22anothernickstestid20210510%22}&start=2021-05-10T01:43:42Z&end=2021-05-10T02:43:42Z&step=300`

**Comment by nickbp** [comment]
Between the two I'd recommend just going with the second one for `query_range`, since the semantics around the timestamps is much clearer with that one. The query is documented by prometheus [here](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries)

**Comment by nickbp** [comment]
note that between logs and metrics, `query_range` takes different timestamp formats for start/end:


- metrics:
> start=<rfc3339 | unix_timestamp>: Start timestamp.
> end=<rfc3339 | unix_timestamp>: End timestamp.

> <rfc3339 | unix_timestamp>: Input timestamps may be provided either in RFC3339 format or as a Unix timestamp in seconds, with optional decimal places for sub-second precision. Output timestamps are always represented as Unix timestamps in seconds.

ref: https://prometheus.io/docs/prometheus/latest/querying/api/#format-overview

- logs:
> start: The start time for the query as a nanosecond Unix epoch. Defaults to one hour ago.
> end: The end time for the query as a nanosecond Unix epoch. Defaults to now.

ref: https://grafana.com/docs/loki/latest/api/#get-lokiapiv1query_range
